pkg/services/ngalert/state/state_tracker.go

```diff
@@ -124,52 +127,52 @@ func (st *StateTracker) ProcessEvalResults(alertRule *ngModels.AlertRule, result
 
 //TODO: When calculating if an alert should not be firing anymore, we should take three things into account:
 // 1. The re-send the delay if any, we don't want to send every firing alert every time, we should have a fixed delay across all alerts to avoid saturating the notification system
-// 2. The evaluation interval defined for this particular alert - we don't support that yet but will eventually allow you to define how often do you want this alert to be evaluted
-// 3. The base interval defined by the scheduler - in the case where #2 is not yet an option we can use the base interval at which every alert runs.
 //Set the current state based on evaluation results
-//return the state and a bool indicating whether a state transition occurred
-func (st *StateTracker) setNextState(alertRule *ngModels.AlertRule, result eval.Result) (AlertState, bool) {
-	currentState := st.getOrCreate(alertRule, result)
+func (st *StateTracker) setNextState(alertRule *ngModels.AlertRule, result eval.Result, processingTime time.Duration) AlertState {
+	currentState := st.getOrCreate(alertRule, result, processingTime)
 	st.Log.Debug("setting alert state", "uid", alertRule.UID)
 	switch {
 	case currentState.State == result.State:
 		st.Log.Debug("no state transition", "cacheId", currentState.CacheId, "state", currentState.State.String())
 		currentState.LastEvaluationTime = result.EvaluatedAt
+		currentState.ProcessingTime = processingTime
 		currentState.Results = append(currentState.Results, StateEvaluation{
 			EvaluationTime:  result.EvaluatedAt,
 			EvaluationState: result.State,
 		})
 		if currentState.State == eval.Alerting {
-			currentState.EndsAt = result.EvaluatedAt.Add(40 * time.Second)
+			currentState.EndsAt = result.EvaluatedAt.Add(alertRule.For * time.Second)
```

Comment:
You should never get to the state cache if the above aren't true. The scheduler and also rule creation logic guards against this. It wouldn't hurt to add it but then it's one more place that would need to get updated if we change that in the future. I think the scheduler is a better place to keep those guards.

Reply:
I don't have a lot of context but I generally agree with @davidmparrott 
But note that the `alertRule.For` (how long the query needs to violate the configured thresholds before the alert notification triggers) can be zero, I'm not sure what should be the `currentState.EndsAt` in that case.
Also, I don't see how this is related with the rule evaluation frequency (this interval [is checked](https://github.com/grafana/grafana/blob/e017de4f061aab133054cf9622392475d80e8d7d/pkg/services/ngalert/store/alert_rule.go#L397) before a rule is saved and [by the scheduler](https://github.com/grafana/grafana/blob/e017de4f061aab133054cf9622392475d80e8d7d/pkg/services/ngalert/schedule/schedule.go#L260))

Reply:
Another comment: Isn't `alertRule.For` already a duration? I don't think we require `* time.Second` (else it's basically `* 1e9`)
(same comment for other places)

Reply:
@papagian handling the case where For is zero is coming in a PR I'm working on tests for now. I haven't pushed it up yet because the tests have revealed a couple bugs in the logic so far and I wanted to not have people spend time reviewing until I'm confident the logic is correct. These state change decisions were all created prior to For being added tot he alert rule and they're not up to date, as of this PR. I can push it up as a branch prior to PRing if anyone is curious though

@codesome you're correct! I'll make that change with the PR I was referring to above

Reply:
https://github.com/grafana/grafana/tree/dparrott/no-data WIP branch